File: zlan6808_allmeth.c
#include <iostream>
#include <string>
#include <cstring>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <bitset>

using namespace std;

class Zlan6808
{
    private:
        int sock;
        string address;
        int port;
        struct sockaddr_in server;
        unsigned char DI1, DI2, DI3, DI4, DI5, DI6, DI7, DI8;
        unsigned char DO1, DO2, DO3, DO4, DO5, DO6, DO7, DO8;

    public:
        Zlan6808(string address, int port)
        {
            this->address = address;
            this->port = port;
            sock = -1;
            memset(&server, 0, sizeof(server));
        }

        bool Connect()
        {
            //Create socket
            sock = socket(AF_INET, SOCK_STREAM, 0);
            if (sock == -1)
            {
                perror("Could not create socket");
                return false;
            }
            cout << "Socket created\n";

            server.sin_addr.s_addr = inet_addr(address.c_str());
            server.sin_family = AF_INET;
            server.sin_port = htons(port);

            //Connect to remote server
            if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0)
            {
                perror("connect failed. Error");
                return false;
            }

            cout << "Connected\n";
            return true;
        }

        /*
        根据传入的字节数组生成CRC16校验码，自动添加在末两字节
        */
        unsigned char* GenCrc16(unsigned char* data, int len)
        {
            unsigned short crc = 0xFFFF;
            for (int i = 0; i < len; i++)
            {
                crc ^= data[i];
                for (int j = 0; j < 8; j++)
                {
                    if ((crc & 0x0001) == 1)
                    {
                        crc >>= 1;
                        crc ^= 0xA001;
                    }
                    else
                    {
                        crc >>= 1;
                    }
                }
            }
            unsigned char* crc_h = new unsigned char[2];
            unsigned char* crc_l = new unsigned char[2];
            crc_h[0] = (unsigned char)((crc & 0xff00) >> 8);
            crc_l[0] = (unsigned char)(crc & 0x00ff);
            unsigned char* crc_res = new unsigned char[2];
            crc_res[0] = crc_l[0];
            crc_res[1] = crc_h[0];
            return crc_res;
        }

        /*
        发送指令接收回包，回包为长度不定的16进制字节串，末两字节为CRC校验码
        */
        unsigned char* SendCmd(unsigned char* cmd, int len)
        {
            if (send(sock, cmd, len, 0) < 0)
            {
                puts("Send failed");
                return NULL;
            }

            unsigned char buffer[1024] = {0};
            int count = recv(sock, buffer, 1024, 0);
            unsigned char* result = new unsigned char[count];
            memcpy(result, buffer, count);
            return result;
        }

        /*
        批量读取DI1到8输入状态，其中第四字节转二进制后，从右至左分别代表DI1到DI8输入状态，1为有输入，2为无输入
        */
        void ReadDi()
        {
            unsigned char data[] = {0x01, 0x01, 0x00, 0x00, 0x00, 0x08};
            int len = sizeof(data)/sizeof(data[0]);
            unsigned char* crc_res = GenCrc16(data, len);
            cout << "CRC:" << hex << (int)crc_res[0] << " " << (int)crc_res[1] << endl;
            unsigned char* repo = SendCmd(crc_res, len+2);
            unsigned char di_byte[] = {repo[6]};
            bitset<8> di_bin(di_byte[0]);
            DI1 = di_bin[7];
            DI2 = di_bin[6];
            DI3 = di_bin[5];
            DI4 = di_bin[4];
            DI5 = di_bin[3];
            DI6 = di_bin[2];
            DI7 = di_bin[1];
            DI8 = di_bin[0];
            cout << "DI状态：" << hex << (int)DI1 << " " << (int)DI2 << " " << (int)DI3 << " " << (int)DI4 << " " << (int)DI5 << " " << (int)DI6 << " " << (int)DI7 << " " << (int)DI8 << endl;
            cout << di_bin << endl;
        }

        /*
        把可阅读的0x00格式命令转换为发送格式
        */
        unsigned char* GenCmd(unsigned char* args, int len)
        {
            return args;
        }

        /*
        设置DO状态
        */
        unsigned char* SetDo(string do_num, string swtich)
        {
            unsigned char data[] = {0x01, 0x05, 0x00, 0x00, 0x00, 0x00};
            int len = sizeof(data)/sizeof(data[0]);
            if (do_num.substr(0, 2) == "DO" && stoi(do_num.substr(2)) >= 1 && stoi(do_num.substr(2)) <= 8)
            {
                data[3] = (unsigned char)(16 + stoi(do_num.substr(2)));
                data[5] = swtich == "on" ? (unsigned char)255 : swtich == "off" ? (unsigned char)0 : throw invalid_argument("开闭指令错误");
            }
            else
            {
                throw invalid_argument("DO接口名错误");
            }
            unsigned char* crc_res = GenCrc16(data, len);
            cout << "CRC:" << hex << (int)crc_res[0] << " " << (int)crc_res[1] << endl;
            unsigned char* repo = SendCmd(crc_res, len+2);
            return repo;
        }
};

int main(int argc, char* argv[])
{
    Zlan6808 zlan6808("192.168.1.200", 4196);
    zlan6808.Connect();
    while (true)
    {
        zlan6808.ReadDi();
    }
    return 0;
}
 char* argv[])
{
    Zlan6808 zlan6808("192.168.1.200", 4196);
    zlan6808.Connect();
    while (true)
    {
        zlan6808.ReadDi();
}

File: zlan6808_allmeth.py
import socket #导入socket模块
import struct

s = socket.socket()	#创建套接字
host = '192.168.1.200' #主机IP
port = 4196
try:
    s.connect((host,port)) #初始化连接
except Exception as e:
    print("连接失败，请检查网络连接或主机IP地址是否正确。")
    print(e)
global DI1,DI2,DI3,DI4,DI5,DI6,DI7,DI8
global DO1,DO2,DO3,DO4,DO5,DO6,DO7,DO8

'''
根据传入的字节数组生成CRC16校验码，自动添加在末两字节
'''
def gen_crc16(data: bytes):
    crc = 0xFFFF
    for i in range(len(data)):
        crc ^= data[i]
        for j in range(8):
            if (crc & 0x0001):
                crc >>= 1
                crc ^= 0xA001
            else:
                crc >>= 1
    crc_h = (crc & 0xff00) >> 8
    crc_l = crc & 0x00ff
    return data + bytes([crc_l, crc_h])

'''
发送指令接收回包，回包为长度不定的16进制字节串，末两字节为CRC校验码
'''
def send_cmd(cmd):
    scmd = struct.pack("%dB" % (len(cmd)), *cmd)
    s.send(scmd)
    print((s.recv(1024)).hex())
    return (s.recv(1024)).hex()

'''
批量读取DI1到8输入状态，其中第四字节转二进制后，从右至左分别代表DI1到DI8输入状态，1为有输入，2为无输入
'''
def read_di():
    data = gen_cmd('0x01','0x01','0x00','0x00','0x00','0x08') #遍历DI指令
    result = gen_crc16(bytes(data))
    print("CRC:",result)
    repo = send_cmd(result)
    di_byte = bytes.fromhex(repo[6:8])
    di_bin = bin(int(di_byte.hex(), 16))[2:].zfill(8)
    DI = [di_bin[i] for i in range(7, -1, -1)]
    print("DI状态：", *DI)
    print(di_bin)

'''
把可阅读的0x00格式命令转换为发送格式
'''
def gen_cmd(*args):
    cmd = []
    for arg in args:
        cmd.append(int(arg, 16))
    return cmd

'''
设置DO状态
'''
def set_do(do, swtich):
    data = gen_cmd('0x01', '0x05', '0x00', '0x00', '0x00', '0x00')
    if do in ['DO1', 'DO2', 'DO3', 'DO4', 'DO5', 'DO6', 'DO7', 'DO8']:
        do_num = int(do[2])
        if swtich == 'on':
            data[3] = 16 + do_num
            data[5] = 255
        elif swtich == 'off':
            data[3] = 16 + do_num
            data[5] = 0
        else:
            print("wrong switch")
    else:
        print("wrong DO")
    result = gen_crc16(bytes(data))
    print("CRC:", result)
    repo = send_cmd(result)
    return repo


'''
循环读取DI状态，测试用
'''
if __name__ == '__main__':
    while True:
            read_di()
